File: Practical_six_sem/AI_Lab/GraphTraversal.java
import java.util.*;

public class GraphTraversal {

    private LinkedList<Integer>[] adj;
    private boolean[] visited;

    
    public GraphTraversal(int v) {
        adj = new LinkedList[v];
        visited = new boolean[v];

        for (int i = 0; i < v; i++) {
            adj[i] = new LinkedList<>();
        }
    }


    void addEdge(int v, int w) {
        adj[v].add(w);
        adj[w].add(v);
    }

    void DFS(int node) {
        visited[node] = true;
        System.out.print(node + " ");

        for (int neighbor : adj[node]) {
            if (!visited[neighbor]) {
                DFS(neighbor);
            }
        }
    }

    void BFS(int start) {
        boolean[] visitedBFS = new boolean[adj.length];
        Queue<Integer> queue = new LinkedList<>();

        visitedBFS[start] = true;
        queue.add(start);

        while (!queue.isEmpty()) {
            int node = queue.poll();
            System.out.print(node + " ");

            for (int neighbor : adj[node]) {
                if (!visitedBFS[neighbor]) {
                    visitedBFS[neighbor] = true;
                    queue.add(neighbor);
                }
            }
        }
    }

    
    void resetVisited() {
        Arrays.fill(visited, false);
    }

  
    public static void main(String[] args) {

        GraphTraversal g = new GraphTraversal(6);

        g.addEdge(0, 1);
        g.addEdge(0, 3);
        g.addEdge(1, 2);
        g.addEdge(2, 4);
        g.addEdge(4, 5);
        g.addEdge(3, 5);

        System.out.println("DFS Traversal:");
        g.DFS(0);

        g.resetVisited();

        System.out.println("\nBFS Traversal:");
        g.BFS(0);
    }
}

File: Engineering/Sem_6/Lab-4/AI/GraphTraversal.py
from collections import deque

class GraphTraversal:
    def __init__(self, v):
        self.adj = [[] for _ in range(v)]
        self.visited = [False] * v

    # Add edge (undirected graph)
    def add_edge(self, v, w):
        self.adj[v].append(w)
        self.adj[w].append(v)

    # DFS (Recursive)
    def DFS(self, node):
        self.visited[node] = True
        print(node, end=" ")

        for neighbor in self.adj[node]:
            if not self.visited[neighbor]:
                self.DFS(neighbor)

    # BFS (Using Queue)
    def BFS(self, start):
        visited_bfs = [False] * len(self.adj)
        queue = deque()

        visited_bfs[start] = True
        queue.append(start)

        while queue:
            node = queue.popleft()
            print(node, end=" ")

            for neighbor in self.adj[node]:
                if not visited_bfs[neighbor]:
                    visited_bfs[neighbor] = True
                    queue.append(neighbor)

    # Reset visited for DFS reuse
    def reset_visited(self):
        self.visited = [False] * len(self.adj)


# Main function
if __name__ == "__main__":
    g = GraphTraversal(6)

    g.add_edge(0, 1)
    g.add_edge(0, 3)
    g.add_edge(1, 2)
    g.add_edge(2, 4)
    g.add_edge(4, 5)
    g.add_edge(3, 5)

    print("DFS Traversal:")
    g.DFS(0)

    g.reset_visited()

    print("\nBFS Traversal:")
    g.BFS(0)